File: src/llamacpp_perfkit/stats.py
from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from math import exp, log, sqrt
from statistics import fmean


def _is_number(value: object) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)


@dataclass(frozen=True)
class Sample:
    values: tuple[float, ...] = ()

    @classmethod
    def from_iterable(cls, values: Iterable[object]) -> Sample:
        return cls(tuple(float(v) for v in values if _is_number(v)))  # type: ignore[arg-type]

    def sorted(self) -> tuple[float, ...]:
        return tuple(sorted(self.values))

    def mean(self) -> float | None:
        return fmean(self.values) if self.values else None

    def median(self) -> float | None:
        if not self.values:
            return None
        ordered = sorted(self.values)
        middle = len(ordered) // 2
        if len(ordered) % 2:
            return ordered[middle]
        return (ordered[middle - 1] + ordered[middle]) / 2

    def stddev(self) -> float | None:
        if len(self.values) < 2:
            return None
        mean = self.mean()
        assert mean is not None
        return sqrt(sum((value - mean) ** 2 for value in self.values) / (len(self.values) - 1))

    def percentile(self, percentile: float) -> float | None:
        if not self.values:
            return None
        ordered = sorted(self.values)
        if len(ordered) == 1:
            return ordered[0]
        bounded = min(100.0, max(0.0, float(percentile)))
        rank = (bounded / 100.0) * (len(ordered) - 1)
        lower = int(rank)
        upper = min(lower + 1, len(ordered) - 1)
        fraction = rank - lower
        return ordered[lower] + (ordered[upper] - ordered[lower]) * fraction

    def geometric_mean(self) -> float | None:
        positives = [value for value in self.values if value > 0]
        if not positives:
            return None
        return exp(sum(log(value) for value in positives) / len(positives))

    def min(self) -> float | None:
        return min(self.values) if self.values else None

    def max(self) -> float | None:
        return max(self.values) if self.values else None


@dataclass(frozen=True)
class MetricSummary:
    count: int = 0
    mean: float | None = None
    median: float | None = None
    p10: float | None = None
    stddev: float | None = None
    geometric_mean: float | None = None
    min: float | None = None
    max: float | None = None

    @classmethod
    def empty(cls) -> MetricSummary:
        return cls()


def summarize(values: Iterable[object]) -> MetricSummary:
    sample = Sample.from_iterable(values)
    return MetricSummary(
        count=len(sample.values),
        mean=sample.mean(),
        median=sample.median(),
        p10=sample.percentile(10),
        stddev=sample.stddev(),
        geometric_mean=sample.geometric_mean(),
        min=sample.min(),
        max=sample.max(),
    )

File: internal/stats/stats.go
package stats

import (
	"math"
	"sort"
)

type MetricSummary struct {
	Count         int      `json:"count"`
	Mean          *float64 `json:"mean"`
	Median        *float64 `json:"median"`
	Stddev        *float64 `json:"stddev"`
	GeometricMean *float64 `json:"geometric_mean"`
	Min           *float64 `json:"min"`
	Max           *float64 `json:"max"`
	P10           *float64 `json:"p10"`
}

func Summarize(values []float64) MetricSummary {
	clean := make([]float64, 0, len(values))
	for _, value := range values {
		if !math.IsNaN(value) && !math.IsInf(value, 0) {
			clean = append(clean, value)
		}
	}
	if len(clean) == 0 {
		return MetricSummary{}
	}

	ordered := append([]float64(nil), clean...)
	sort.Float64s(ordered)

	mean := average(clean)
	median := percentileSorted(ordered, 50)
	p10 := percentileSorted(ordered, 10)
	minVal := ordered[0]
	maxVal := ordered[len(ordered)-1]
	geo := geometricMean(clean)

	out := MetricSummary{
		Count:  len(clean),
		Mean:   &mean,
		Median: &median,
		Min:    &minVal,
		Max:    &maxVal,
		P10:    &p10,
	}
	if len(clean) > 1 {
		std := sampleStddev(clean, mean)
		out.Stddev = &std
	}
	if geo != nil {
		out.GeometricMean = geo
	}
	return out
}

func average(values []float64) float64 {
	var sum float64
	for _, value := range values {
		sum += value
	}
	return sum / float64(len(values))
}

func sampleStddev(values []float64, mean float64) float64 {
	var sum float64
	for _, value := range values {
		diff := value - mean
		sum += diff * diff
	}
	return math.Sqrt(sum / float64(len(values)-1))
}

func geometricMean(values []float64) *float64 {
	var sum float64
	var count int
	for _, value := range values {
		if value <= 0 {
			continue
		}
		sum += math.Log(value)
		count++
	}
	if count == 0 {
		return nil
	}
	out := math.Exp(sum / float64(count))
	return &out
}

func percentileSorted(ordered []float64, pct float64) float64 {
	if len(ordered) == 1 {
		return ordered[0]
	}
	bounded := math.Max(0, math.Min(100, pct))
	rank := (bounded / 100) * float64(len(ordered)-1)
	lower := int(rank)
	upper := lower + 1
	if upper >= len(ordered) {
		upper = len(ordered) - 1
	}
	fraction := rank - float64(lower)
	return ordered[lower] + (ordered[upper]-ordered[lower])*fraction
}
